pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,606 @@
+		if key, err := controller.KeyFunc(change.oldService); err == nil {
+			ic.clearLocalAllocation(key, oldIP)
+		} else {
+			// Recovery/retry not possible for this error
+			utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", change.oldService, err))
+		}
+	case service != nil && service.Spec.Type == kapi.ServiceTypeLoadBalancer && len(service.Status.LoadBalancer.Ingress) > 0:
+		// Service with existing allocation
+		return ic.recordAllocation(service, change.key, oldIP)
+	case service != nil && service.Spec.Type == kapi.ServiceTypeLoadBalancer:
+		// Service requiring allocation
+		return ic.allocate(service, change.key) // TODO is the oldip relevant here?
+	case service != nil && len(service.Status.LoadBalancer.Ingress) > 0:
+		// A service for which allocation status exists but that is no
+		// longer of type LoadBalancer should be deallocated.
+		return ic.deallocate(service, change.key) // TODO is the oldip relevant here?
+	}
+
+	return nil
+}
+
+// recordAllocation updates local state with the ingress ip indicated
+// in a service's status and ensures that the ingress ip appears in
+// the service's list of external ips.
+func (ic *IngressIPController) recordAllocation(service *kapi.Service, key, oldIP string) error {
+	// If more than one ingress ip is present, it will be ignored
+	ipString := service.Status.LoadBalancer.Ingress[0].IP
+
+	// Clear a previously allocated ip
+	if len(oldIP) > 0 && ipString != oldIP {
+		ic.clearLocalAllocation(key, oldIP)
+	}
+
+	reallocate, err := ic.recordLocalAllocation(key, ipString)
+
+	if reallocate || err == nil {
+		// Make a copy to modify to avoid mutating cache state
+		t, err := kapi.Scheme.DeepCopy(service)
+		if err != nil {
+			return err
+		}
+		service = t.(*kapi.Service)
+	}
+
+	switch {
+	case reallocate:
+		return ic.reallocate(service, key, err.Error())
+	case err == nil:
+		// Ensure that the ingress ip is present in the service's spec.
+		return ic.ensureExternalIP(service, key, ipString)
+	default:
+		return err
+	}
+}
+
+func (ic *IngressIPController) reallocate(service *kapi.Service, key, errMsg string) error {
+	// TODO update the external ips but not the status since
+	// allocate() will overwrite any existing allocation.
+	if err := ic.clearPersistedAllocation(service, key, errMsg); err != nil {
+		return err
+	}
+	return ic.allocate(service, key)
+}
+
+// allocate assigns an unallocated ip to a service and updates the
+// service's persisted state.
+func (ic *IngressIPController) allocate(service *kapi.Service, key string) error {
+	// Make a copy to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(service)
+	if err != nil {
+		return err
+	}
+	service = t.(*kapi.Service)
+
+	ip, err := ic.allocateIP(service.Spec.LoadBalancerIP)
+	if err != nil {
+		return err
+	}
+	ipString := ip.String()
+
+	glog.V(5).Infof("Allocating ip %v to service %v", ipString, key)
+	service.Status = kapi.ServiceStatus{
```

Comment:
allocate() is only going to be called if:

 - a service is of type load balancer but does not have an ingress ip in its status
 - an existing allocation record is invalid (bad ip, ip out of range, ip in use)
 
In the latter case, I've been assuming that automatic reallocation is required.  Is your comment intended to suggest another option?